src/main/kotlin/persistence/repo/UserRepo.kt

```diff
@@ -0,0 +1,132 @@
+import persistence.mapping.UserMapper
+import persistence.model.IUserEntity
+import persistence.model.IUserLanguage
+import persistence.model.UserLanguage
+
+class UserRepo(private val dataStore: KotlinEntityDataStore<Persistable>,
+               private val userLanguageRepo: UserLanguageRepo, languageDao : Dao<Language>): Dao<User> {
+    private val userMapper = UserMapper(userLanguageRepo, languageDao)
+    /**
+     * function to create and insert a user into the database
+     * takes in a audioHash and a path to a recording to creaete
+     */
+    override fun insert(user: User): Observable<Int> {
+        // creates observable to return generated int
+        return Observable.create<Int> {
+            it.onNext(dataStore.insert(userMapper.mapToEntity(user)).id)
+        }.doOnNext {
+            updateUserLanguageReferences(user, it)
+        }.subscribeOn(Schedulers.io())
+    }
+
+    /**
+     * gets user by Id
+     */
+    override fun getById(id:Int): Observable<User>{
+        return Observable.create<IUserEntity> {
+            it.onNext(dataStore.select(IUserEntity::class)
+                    .where(IUserEntity::id eq id)
+                    .get().first())
+        }.map {
+            userMapper.mapFromEntity(it)
+        }.subscribeOn(Schedulers.io())
+
+    }
+
+    /**
+     * given a audioHash gets the user
+     */
+    fun getByHash(hash: String): Observable<User> {
+        return Observable.create<IUserEntity> {
+            dataStore.select(IUserEntity::class)
+                    .where(IUserEntity::audioHash eq hash)
+                    .get().first()
+        }.map {
+            userMapper.mapFromEntity(it)
+        }.subscribeOn(Schedulers.io())
+    }
+
+    /**
+     * gets all the users currently stored in db
+     */
+    override fun getAll(): Observable<List<User>>{
+        return Observable.create<List<IUserEntity>> {
+            it.onNext(dataStore.select(IUserEntity::class)
+                    .get().toList())
+        }.map {
+            it.map { userMapper.mapFromEntity(it) }
+        }.subscribeOn(Schedulers.io())
+
+    }
+
+    override fun update(user: User): Completable {
+        return Completable.fromAction {
+                val userEntity = userMapper.mapToEntity(user)
+                dataStore.update(userEntity)
+                dataStore.update(userEntity.userPreferencesEntity)
+                updateUserLanguageReferences(user, user.id)
+        }.subscribeOn(Schedulers.io())
+    }
+
+    /**
+     * deletes user by id
+     */
+    override fun delete(user: User): Completable{
+        return Completable.fromAction {
+            val userEntity = userMapper.mapToEntity(user)
+            dataStore.delete(userEntity)
+            dataStore.delete(IUserLanguage::class).where(IUserLanguage::userEntityid eq user.id).get().value()
+        }.subscribeOn(Schedulers.io())
+    }
+
+    private fun updateUserLanguageReferences(user: User, userId: Int) {
```

Comment:
might want to consider changing this function, otherwise to add new source and target languages, a new User has to be created with a new updated list but the same id path and hash which could lead to errors. Maybe create a function that takes a language or list of languages, a user, and whether its adding to source or not then updates the database

Reply:
Probably can be removed in an upcoming fix